compat/mingw.c

```diff
@@ -426,6 +426,67 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+static BOOL exists(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_READ,
+				 FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+				 OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return FALSE;
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+static BOOL is_exclusively_writable(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return !is_file_in_use_error(GetLastError());
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *wpathname, int (*unlinker)(const wchar_t *))
+{
+	wchar_t wtemppath[MAX_LONG_PATH];
+	wchar_t *last, *original;
+	*wtemppath = 0;
+	/* leave some space for the .unlink extension */
+	wcsncat(wtemppath, wpathname, sizeof(wtemppath) - 8);
+	original = last = wtemppath + wcslen(wtemppath);
+	/* Start with file.unlink. If it exists, append X up to 12 times */
+	wcscpy(last, L".unlink");
+	last += (sizeof(L".unlink") / sizeof (wchar_t)) - 1;
+	while (exists(wtemppath)) {
+		if ((last - original >= 20) || (last - wtemppath + 1 >= MAX_LONG_PATH)) {
+			last = original;
+			*last = 0;
+			break;
+		}
+		*last++ = L'X';
+		*last = 0;
+	}
+	if (last > original) {
+		int res = 0;
+		if (MoveFileExW(wpathname, wtemppath, 0)) {
+			res = unlinker(wtemppath);
+			if (res)
+				MoveFileExW(wtemppath, wpathname, 0);
```

Comment:
I'd rather not take emergency actions if we don't really know what's going on.

Reply:
That's not emergency, it just renames the file back, then another attempt is made with rmdir, and then 4 more iterations before giving up.

Reply:
If the rename back fails then the retry makes no sense either. Do it right or don't do it at all.

Reply:
The current logic in unlink first calls unlink, and if it falls, it calls rmdir. If we don't rename back, rmdir will be called for a non-existent file. This broke checkout of symbolic links for directories.